File: lib/llm/hybrid.mjs
// lib/llm/hybrid.mjs
// Hybrid LLM Provider — routes tasks to different models strategically
//
// DeepSeek: complex reasoning, brain assessments, strategic analysis (best quality)
// ARIA Local (Ollama): fast queries, knowledge lookups, simple Q&A (lowest cost, fastest)
//
// Config:
//   LLM_PROVIDER=hybrid
//   LLM_API_KEY=deepseek-api-key
//   OLLAMA_URL=http://localhost:11434  (or remote GPU server)
//   OLLAMA_MODEL=aria-v1              (your fine-tuned model)
//
// Routing logic:
//   - Brain assessments → DeepSeek (needs best reasoning)
//   - Think mode (6-step analysis) → DeepSeek
//   - ARIA chat (simple Q&A) → Ollama first, DeepSeek fallback
//   - Ideas generation → DeepSeek
//   - Everything else → Ollama first, DeepSeek fallback

import { LLMProvider } from './provider.mjs';
import { OllamaProvider } from './ollama.mjs';
import { OpenAIProvider } from './openai.mjs';

export class HybridProvider extends LLMProvider {
  constructor(config) {
    super(config);
    this.name = 'hybrid';

    // Primary: DeepSeek (best quality, costs money)
    this.primary = new OpenAIProvider({
      apiKey: config.apiKey,
      model: config.model || 'deepseek-chat',
      baseUrl: config.baseUrl || 'https://api.deepseek.com/v1',
    });

    // Secondary: Ollama (local/self-hosted, free, fast)
    this.secondary = new OllamaProvider({
      model: config.ollamaModel || process.env.OLLAMA_MODEL || 'llama3.1:8b',
      baseUrl: config.ollamaUrl || process.env.OLLAMA_URL || 'http://localhost:11434',
    });

    // Track if Ollama is reachable
    this._ollamaAvailable = null;
    this._lastOllamaCheck = 0;
  }

  get isConfigured() {
    return this.primary.isConfigured;
  }

  /**
   * Check if Ollama is reachable (cached for 5 minutes).
   */
  async _isOllamaReady() {
    const now = Date.now();
    if (this._ollamaAvailable !== null && now - this._lastOllamaCheck < 300000) {
      return this._ollamaAvailable;
    }
    try {
      const r = await fetch(this.secondary.baseUrl + '/api/tags', {
        signal: AbortSignal.timeout(3000),
      });
      this._ollamaAvailable = r.ok;
    } catch {
      this._ollamaAvailable = false;
    }
    this._lastOllamaCheck = now;
    return this._ollamaAvailable;
  }

  /**
   * Route to appropriate model based on task complexity.
   */
  async complete(systemPrompt, userMessage, opts = {}) {
    const useDeepSeek = this._shouldUseDeepSeek(systemPrompt, userMessage, opts);

    if (useDeepSeek) {
      // Complex task → DeepSeek (best quality)
      try {
        const result = await this.primary.complete(systemPrompt, userMessage, opts);
        return { ...result, _routed: 'deepseek' };
      } catch (err) {
        console.warn('[Hybrid] DeepSeek failed, trying Ollama fallback:', err.message);
        // Fall through to Ollama
      }
    }

    // Simple task or DeepSeek failed → try Ollama first
    const ollamaReady = await this._isOllamaReady();
    if (ollamaReady) {
      try {
        const result = await this.secondary.complete(systemPrompt, userMessage, opts);
        return { ...result, _routed: 'ollama' };
      } catch (err) {
        console.warn('[Hybrid] Ollama failed:', err.message);
      }
    }

    // Ollama not available or failed → DeepSeek as final fallback
    const result = await this.primary.complete(systemPrompt, userMessage, opts);
    return { ...result, _routed: 'deepseek-fallback' };
  }

  /**
   * Determine if this task needs DeepSeek (complex) or can use Ollama (simple).
   */
  _shouldUseDeepSeek(systemPrompt, userMessage, opts) {
    const sys = (systemPrompt || '').toLowerCase();
    const msg = (userMessage || '').toLowerCase();

    // Brain assessment — always DeepSeek (needs best reasoning + JSON output)
    if (sys.includes('autonomous bd intelligence brain') || sys.includes('crucix')) return true;

    // Think mode — always DeepSeek (6-step deep analysis)
    if (sys.includes('deep reasoning protocol') || sys.includes('6-step')) return true;

    // Ideas generation — always DeepSeek
    if (msg.includes('generate') && (msg.includes('ideas') || msg.includes('strategy'))) return true;

    // Long response needed — DeepSeek handles better
    if (opts.maxTokens > 2000) return true;

    // Complex queries (compliance, multi-market, financial)
    if (msg.includes('compliance') || msg.includes('itar') || msg.includes('export control')) return true;
    if (msg.includes('compare') || msg.includes('versus') || msg.includes('which market')) return true;

    // Everything else → Ollama (fast, free)
    return false;
  }
}

File: aria_service/llm/hybrid.py
"""
Hybrid provider — routes between DeepSeek (complex) and Ollama (simple).
Falls back gracefully if one provider is unavailable.
"""
from __future__ import annotations

import logging
import time
from .provider import LLMProvider, LLMResult
from .openai_compat import OpenAICompatProvider
from ..intel.wire import fail_wire  # R-F1789 §21 brain-wiring

logger = logging.getLogger("aria.llm.hybrid")


class HybridProvider(LLMProvider):
    name = "hybrid"

    def __init__(
        self,
        *,
        deepseek_key: str,
        deepseek_model: str = "deepseek-chat",
        deepseek_url: str = "https://api.deepseek.com/v1",
        ollama_model: str = "llama3.1:8b",
        ollama_url: str = "http://localhost:11434",
    ):
        self.primary = OpenAICompatProvider(
            name="deepseek",
            api_key=deepseek_key,
            model=deepseek_model,
            base_url=deepseek_url,
        )
        self.secondary = OpenAICompatProvider(
            name="ollama",
            api_key="",
            model=ollama_model,
            base_url=f"{ollama_url.rstrip('/')}/v1",
            default_timeout=120.0,
        )
        self._ollama_ok: bool | None = None
        self._ollama_checked: float = 0
        self.stats = {
            "deepseek_calls": 0,
            "ollama_calls": 0,
            "fallback_calls": 0,
            "deepseek_tokens": 0,
            "ollama_tokens": 0,
            "started_at": time.time(),
        }

    @property
    def is_configured(self) -> bool:
        return self.primary.is_configured

    async def _is_ollama_ready(self) -> bool:
        now = time.time()
        if self._ollama_ok is not None and now - self._ollama_checked < 300:
            return self._ollama_ok
        import httpx
        try:
            async with httpx.AsyncClient(timeout=3.0) as c:
                r = await c.get(f"{self.secondary._base_url.replace('/v1','')}/api/tags")
                self._ollama_ok = r.is_success
        except Exception:
            self._ollama_ok = False
        self._ollama_checked = now
        return self._ollama_ok

    def _should_use_deepseek(self, system: str, user: str, max_tokens: int) -> bool:
        sl = system.lower()
        ul = user.lower()
        if "autonomous bd intelligence brain" in sl or "crucix" in sl:
            return True
        if "deep reasoning protocol" in sl or "6-step" in sl:
            return True
        if "generate" in ul and ("ideas" in ul or "strategy" in ul):
            return True
        if max_tokens > 2000:
            return True
        if any(k in ul for k in ("compliance", "itar", "export control", "compare", "versus", "which market")):
            return True
        return False

    @fail_wire(module="hybrid", gap_type="engine_failure")
    async def complete(
        self,
        system_prompt: str,
        user_message: str,
        *,
        max_tokens: int = 4096,
        timeout: float = 60.0,
    ) -> LLMResult:
        use_ds = self._should_use_deepseek(system_prompt, user_message, max_tokens)

        if use_ds:
            try:
                r = await self.primary.complete(system_prompt, user_message, max_tokens=max_tokens, timeout=timeout)
                self.stats["deepseek_calls"] += 1
                self.stats["deepseek_tokens"] += r.input_tokens + r.output_tokens
                r.routed_via = "deepseek"
                return r
            except Exception as e:
                logger.warning(f"DeepSeek failed, trying Ollama: {e}")

        if await self._is_ollama_ready():
            try:
                r = await self.secondary.complete(system_prompt, user_message, max_tokens=max_tokens, timeout=120.0)
                self.stats["ollama_calls"] += 1
                self.stats["ollama_tokens"] += r.input_tokens + r.output_tokens
                r.routed_via = "ollama"
                return r
            except Exception as e:
                logger.warning(f"Ollama failed: {e}")

        # Final fallback
        r = await self.primary.complete(system_prompt, user_message, max_tokens=max_tokens, timeout=timeout)
        self.stats["fallback_calls"] += 1
        self.stats["deepseek_tokens"] += r.input_tokens + r.output_tokens
        r.routed_via = "deepseek-fallback"
        return r

    @fail_wire(module="hybrid", gap_type="engine_failure")
    def get_stats(self) -> dict:
        return {
            **self.stats,
            "total_calls": self.stats["deepseek_calls"] + self.stats["ollama_calls"] + self.stats["fallback_calls"],
        }
